### PR TITLE
 CA-287921: nbd_client_manager: track nbd device -> nbd server mapping 

### PR DESCRIPTION
**Pull request by gaborigloi**
Merge with https://github.com/xapi-project/vhd-tool/pull/56

Keep track of the mapping of the connected /dev/nbdX NBD devices to the
Unix domain socket of the server and the exportname used for the
connection. This allows sparse_dd to look up the NBD connection details
based on the NBD device it receives, then it can extract the list of
allocated extents to copy from the NBD server using the BLOCK_STATUS NBD
extension.

Also, add randomized quicktest to check VDI.copy data integrity

Signed-off-by: Gabor Igloi <gabor.igloi@citrix.com>

### CODE REVIEW

**Comment by gaborigloi** [comment]
Merge with https://github.com/xapi-project/vhd-tool/pull/56

File: ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+let fill rpc session_id vdi =
+  let size =
+    Client.Client.VDI.get_virtual_size ~rpc ~session_id ~self:vdi
+    |> Int64.to_int
+  in
+  with_open_vdi rpc session_id vdi `RW
+    (fun fd ->
+       let buf = Bytes.make size 'a' in
+       Unix.write fd buf 0 size
+    )
```
Comment (gaborigloi) [inline ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: Probably I should fill the VDI with more random data, instead of using the same char everywhere. I just couldn't find a library that gives me a random `bytes` of a given length

File: ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  with_open_vdi rpc session_id vdi `RW
+    (fun fd ->
+
+       let max_writes = 100 in
+       let max_write_length = 2048 in
+
+       let write_random_char_block offset length =
+         let buf =
+           let c = Random.int 256 |> Char.chr in
+           Bytes.make length c
```
Comment (gaborigloi) [inline ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: Same comment as below - ideally this would be more random than just using the same char.

File: ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml (hunk 1 above)
Comment (mseri) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: I know only https://github.com/xavierleroy/cryptokit/blob/master/src/cryptokit.mli#L241

Comment (gaborigloi) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: Or we could copy from `/dev/urandom`?

Comment (mseri) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: If you look at the various implementations, that's what one of the options in cryptokit does :D
I think we can even use the Random module and fill the buffer.

Comment (edwintorok) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: Generating random numbers with a PRNG can be faster than `/dev/urandom` (especially if you just use something fast like a shift register, worse quality randomness but fast), however using `String.init` with `Random.int` is quite slow:
```
String.init Random: 0.225s
/dev/urandom: 0.045s
```

```
let time msg f =
  let t0 = Unix.gettimeofday () in
  ignore(f ());
  let t1 = Unix.gettimeofday () in
  Printf.printf "%s: %.3fs\n" msg (t1 -. t0)

let () =
  Random.self_init ();
  let n = 10_000_000 in

  time "String.init Random" (fun () ->
      String.init n (fun _ -> Char.chr (Random.int 255)));

  time "/dev/urandom" (fun () ->
      let f = open_in "/dev/urandom" in
      let r = really_input_string f n in
      close_in f;
      r)
```

A very fast and simple PRNG with reasonable quality is 2CMRES from here:
http://www.drdobbs.com/tools/fast-high-quality-parallel-random-number/231000484?pgno=2
This one should also be quite fast: http://www.pcg-random.org/
As long as you don't need it for cryptographic purposes either of these algorithms should work well.

Comment (gaborigloi) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: Thanks for the investigation, I think I'll go for `/dev/urandom` for now. There is also [mirage-random](https://github.com/mirage/mirage-random) in xs-opam, but it uses `Random.int`.

File: ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+      check_vdi_unchanged test_nodata rpc session_id sr
+        ~prepare_vdi:(fun _vdi -> ());
+
+      let test_random = Quicktest_common.make_test "Copy of random VDI" 6 in
+      check_vdi_unchanged test_random rpc session_id sr
+        ~prepare_vdi:(write_random_data rpc session_id);
+
+      let test_full = Quicktest_common.make_test "Copy of full VDI" 6 in
+      check_vdi_unchanged test_full rpc session_id sr
+        ~prepare_vdi:(fill rpc session_id)
```
Comment (edwintorok) [inline ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: This is a good start. Might be useful to explicitly cover a few edge cases: hole at beginning of file (multiple of 64k), hole at end of file (multiple of 64k), write some data, then write random multiple of 64k zeroes (don't know if qemu would convert the zeroes to sparse, it might).

Also how do snapshots and leaf-coalesce affect this functionality? Might be useful to add a few tests that take write some random data with holes, take a snapshot, modify the original vdi again, check export of original vdi/snapshotted vdi, then delete snapshot, and do some copies again while SMGC is running. Although if there are any bugs there they would be in qemu/SMGC, so perhaps such a test is better written as part of the pool stress test. Could you verify if there are any such tests in the pool stress, and if not create a ticket to add some?

Comment (gaborigloi) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: Why should we write multiples of 64K blocks in the test? In my opinion completely random block lengths / offsets cover more edge cases. I could add another test case though for writing zeroes into the VDI.
I agree that those more complex scenarios should be run by XenRT. I'll check what VDI.copy tests we have at the moment.

Comment (edwintorok) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: Right, I think that you need a minimum of 64k, not necesarily a multiple of 64k. So 64k+random would cover more edge cases true.

Comment (gaborigloi) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: Not sure how we could detect this from a quicktest, maybe by checking the `physical_utilization` field if that's reliable. But there is a XenRT test that was failing because the copy did not took advantage of sparseness.

Comment (edwintorok) [reply ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml]: The VDI size is 4M, hence there is a reasonable chance that we will have >64k in our test. More tests could be added in a separate PR.
